src/cargo/ops/cargo_install.rs

```diff
@@ -85,15 +566,38 @@ pub fn install(
                 no_track,
                 !did_update,
             ) {
-                Ok(still_needs_update) => {
+                Ok(Some(installable_pkg)) => {
+                    did_update = true;
+                    Some((krate, installable_pkg))
+                }
+                Ok(None) => {
+                    // Already installed
                     succeeded.push(krate);
-                    did_update |= !still_needs_update;
+                    None
                 }
                 Err(e) => {
                     crate::display_error(&e, &mut config.shell());
                     failed.push(krate);
                     // We assume an update was performed if we got an error.
                     did_update = true;
+                    None
+                }
+            }
+        });
+
+        let install_results: Vec<_> = pkgs_to_install
+            .into_iter()
+            .map(|(krate, installable_pkg)| (krate, installable_pkg.install_one()))
+            .collect();
+
+        for (krate, result) in install_results {
+            match result {
+                Ok(()) => {
+                    succeeded.push(krate);
+                }
+                Err(e) => {
+                    crate::display_error(&e, &mut config.shell());
```

Comment:
This seems like we may want to handle errors earlier perhaps? Otherwise since we process all the packages I think this could make it somewhat difficult to associated packages with errors, and it may also run the risk of delaying errors for some time while other packages are being processed?

I'm not entirely certain, though, what sort of errors are happening during the selection of packages, so this may be more reasonable than I expect. I think, though, that this is a behavior change since today it's more of a fail-fast situation, right?

Reply:
Ideally errors happening during selection will show up next to "Downloading crates" and errors happening during compilation will happen next to "Compiling foo", so it's clear in either case.

Once I add the collect above - we can see the test output.

Reply:
Ah ok I think I also missed that this is actually roughly what happened earlier, sorry I got lost in the diff!